scipy/stats/_continuous_distns.py

```diff
@@ -8420,6 +8420,270 @@ def _fitstart(self, data):
+                if not res.converged:
+                    return fallback(data, *args, **kwargs)
+
+                # Determine the MLE for loc.
+                # Iteratively look for a bracket for root_scalar.
+                rbrack = res.root - 1e-3  # grad_loc is numerically ill-behaved
+                lbrack = rbrack - 1
+                i = 0
+                while ((lbrack > -np.inf)
+                       and (dL_dLoc(lbrack)*dL_dLoc(rbrack) >= 0)):
+                    i += 1
+                    lbrack = rbrack - np.power(2., i)
+                if not lbrack > -np.inf:
+                    return fallback(data, *args, **kwargs)
+                res = root_scalar(dL_dLoc, bracket=(lbrack, rbrack))
+                if not res.converged:
+                    return fallback(data, *args, **kwargs)
+                loc = res.root
+                shape_b = get_b(loc)
+                shape_c = get_c(loc)
+                scale = get_scale(loc)
+
+                std_data = (data - loc)/scale
+                # The expression of b relies on b being bounded above.
+                up_bound_b = min(1/log_mean(std_data),
+                                 1/(harm_mean(std_data)-1))
+                if not (shape_b < up_bound_b):
+                    return fallback(data, *args, **kwargs)
+            else:
+                # We know b is positive (or a FitError will be triggered)
+                # so we let loc get close to min(data).
+                rbrack = mn_inf
+                lbrack = mn_inf - 1
+                i = 0
+                # Iteratively look for a bracket for root_scalar.
+                while (lbrack > -np.inf
+                       and (dL_dLoc(lbrack, fshape_b)
+                            * dL_dLoc(rbrack, fshape_b) >= 0)):
+                    i += 1
+                    lbrack = rbrack - 2**i
+                if not lbrack > -np.inf:
+                    return fallback(data, *args, **kwargs)
+                res = root_scalar(dL_dLoc, (fshape_b,),
+                                  bracket=(lbrack, rbrack))
+                if not res.converged:
+                    return fallback(data, *args, **kwargs)
+                loc = res.root
+                shape_c = get_c(loc)
+                scale = get_scale(loc)
+                shape_b = fshape_b
+        else:
+            # At least one of the parameters determining the support is fixed;
+            # the others then have analytical expressions from the constraints.
+            # Overdetermined cases (fixed c and either loc or scale) behave.
+            # The completely determined case (fixed c, loc and scale)
+            # has to be checked for not overflowing the support.
+            # If not fixed, b has to be determined numerically.
+
+            # To make the following switch structure more readable,
+            # the "fixedness" of the parameters is associated an octal digit.
+            bm = (4*(fshape_c is not None)
```

Comment:
This is a bit matlabish or C-ish programming by the lack of proper data structure but you can clean this up by providing a single assignment since None checks are very convenient in Python and also you are using nonlocal variables in `get_xxx_010` type func calls which requires unnecessary care.  If 000 and 111 cases are handled prior to this point then you can easily go with 

```python
shape_c = fshape_c if fshape_c else compute_shape(floc=floc, fscale=fscale)
loc = floc if floc else compute_loc(fshape=fshape, fscale=fscale)
scale = fscale if fscale else compute_scale(fshape=fshape, floc=floc)
```

Then you can also consolidate the different options in one helper function `compute_shape(args)` etc. That would keep things together and much easier to spot the options/bugs and so on. Currently they are all spread out and difficult to concentrate to follow the flow.